File: typescript/packages/bench/src/concerns/hasFormatting.ts
/**
 * Structural concern.
 *
 * No class declares `implements HasFormatting`: any object with the two methods
 * satisfies it structurally, which is the TypeScript analogue of a blanket impl.
 */

export interface HasFormatting {
    formatValue(value: number): string;
    formatLabel(label: string): string;
}

/** The shared body every component reuses. */
export const formatting: HasFormatting = {
    /** Two-decimal rendering used by every report footer. */
    formatValue(value: number): string {
        return value.toFixed(2);
    },

    formatLabel(label: string): string {
        return label.trim();
    },
};

/** Structural check: does `candidate` satisfy the concern at runtime? */
export function isFormatting(candidate: unknown): candidate is HasFormatting {
    return (
        typeof candidate === "object" &&
        candidate !== null &&
        typeof (candidate as HasFormatting).formatValue === "function"
    );
}

File: rust/crates/atelier-bench/src/concerns/has_formatting.rs
//! Blanket concern.
//!
//! Every [`ReportContract`] implementor gets these methods without naming this
//! trait: a `format_value` call site has no impl block to jump to, only the
//! blanket impl below.

use crate::contracts::report_contract::ReportContract;

pub trait HasFormatting {
    /// Two-decimal rendering used by every report footer.
    fn format_value(&self, value: f64) -> String {
        format!("{value:.2}")
    }

    fn format_label(&self, label: &str) -> String {
        label.trim().to_string()
    }
}

impl<T: ReportContract> HasFormatting for T {}
